### PR TITLE
fix: Persist data

### LINKED ISSUE #109: bug: App makes network calls everytime user changes page
**Is your feature request related to a problem? Please describe.**
Right now every time we switch pages, the app makes a network call and gets the data, this makes the app feel slow. 

**Describe the solution you'd like**
Don't get data from backend until the user refreshes. User refresh action can be set ti pull down to refresh

Right now (App loafs page every time from backend):

test

### PR DESCRIPTION
**Pull request by techno-disaster**
### Description

Fixes #109 

Right now every time we switch pages, the app makes a network call and gets the data, this makes the app feel slow. With this fix the app persists the data on the page and is updated when user pulls down to refresh. This is similar to the store data locally PR on mentorship-android [here](https://github.com/anitab-org/mentorship-android/pull/718#) , but more developed because 

1. Does not store it in local storage,(not exactly), the data is persisted in the state itself by blocs and app does not make new network calls untill pulled down to refreshed or updated from something in the page itself like updating the profile or creating/deleting tasks.

2. Its for all the pages not only profile page data. So if someone makes a change on the web version (in development) or ios, just one pull down to refresh and the changes are updated on screen. 

### Flutter Channel:
- [x] I have used the Flutter Beta channel on my local machine 

### Type of Change:
**Delete irrelevant options.**

- Code
- User Interface

**Code/Quality Assurance Only**

- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Physical device.
PS: videos wont show how much improvment this actually brings because it is capped at 7 Fps. i would suggest to test this yourself manually on your device/emulator

|Before|After|
|---|---|
| before|after  |

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by techno-disaster** [comment]
@robotjellyzone feel free to test this :) 
almost the whole app state has been changed, so you might have to check all the pages

File: lib/screens/home/pages/profile/bloc/profile_page_bloc.dart [hunk 1]
```diff
@@ -31,6 +41,21 @@ class ProfilePageBloc extends Bloc<ProfilePageEvent, ProfilePageState> {
...
+  }
+
+  Stream<ProfilePageState> mapEventToRefreshRequested(ProfilePageEvent event) async* {
+    if (event is ProfilePageRefresh) {
+      yield ProfilePageLoading();
+      try {
+        _user = await userRepository.getCurrentUser();
+        yield ProfilePageSuccess(_user, message: event.message);
+      } on Failure catch (_) {
+        yield state;
```
Comment (bartekpacia) [inline lib/screens/home/pages/profile/bloc/profile_page_bloc.dart:54]: Add `Logger.root.error` or something like this

File: lib/screens/home/pages/members/members_page.dart [hunk 2]
```diff
@@ -1,74 +1,82 @@
...
+        _refreshCompleter?.complete();
+        _refreshCompleter = Completer();
+      }
+    }, builder: (context, state) {
+      return BlocBuilder<MembersPageBloc, MembersPageState>(
+        builder: (context, state) {
+          if (state is MembersPageFailure) {
             return Center(
-              child: Text('no users'),
+              child: Text('failed to get users'),
```
Comment (bartekpacia) [inline lib/screens/home/pages/members/members_page.dart]: first uppercase?

**Review by bartekpacia — DISMISSED** [review]
good job, take a look at my nitpicky comments ;P

File: lib/screens/home/pages/profile/bloc/profile_page_bloc.dart (hunk 1 above)
Comment (techno-disaster) [reply lib/screens/home/pages/profile/bloc/profile_page_bloc.dart:54]: Done

File: lib/screens/home/pages/members/members_page.dart (hunk 2 above)
Comment (techno-disaster) [reply lib/screens/home/pages/members/members_page.dart]: Done

File: lib/screens/home/pages/requests/bloc/requests_page_bloc.dart [hunk 3]
```diff
@@ -45,7 +45,8 @@ class RequestsPageBloc extends Bloc<RequestsPageEvent, RequestsPageState> {
       try {
         List<Relation> relations = await relationRepository.getAllRelationsAndRequests();
         yield RequestsPageSuccess(relations);
-      } on Failure catch (_) {
+      } on Failure catch (failure) {
+        Logger.root.severe(failure.message);
```
Comment (bartekpacia) [inline lib/screens/home/pages/requests/bloc/requests_page_bloc.dart]: `Logger.root.severe(failure.message);` -> `Logger.root.severe("RequestsPageBloc: Failure catched: failure.message`);`

Do this in similar places too^

We need to find a better way to log stuff but for now, let's just do logging like I've shown above

Comment (techno-disaster) [reply lib/screens/home/pages/requests/bloc/requests_page_bloc.dart]: done

**Review by bartekpacia — DISMISSED** [review]
great:D

**Comment by techno-disaster** [comment]
@anitab-org/qa-team can someone test this?  almost the whole app state has been changed, so you might have to check all the pages

**Comment by rpattath** [comment]
@anitab-org/mentorship-flutter-maintainers please provide one more approval so that QA team can start testing.

**Review by HaripriyaB — DISMISSED** [review]

**Comment by techno-disaster** [comment]
@isabelcosta Can you check this ones? Merging this would greatly improve user experience

**Comment by HaripriyaB** [comment]
Awesome @Techno-Disaster !
The changes made in this PR were tested locally. Following are the results:

1. Code review - Done

2. All possible responses (positive and negative tests) were tested as below:

  * Each page was tested    
    Screenshot/gif:  
Record-2020-06-30-23-20-39

    _Expected Result_: The pages load for the first time and later on it doesn't load when navigated through the pages. 
    _Actual Result_: As expected!

3. Status of PR Changed to: Ready to Merge ✔️

**Comment by techno-disaster** [comment]
@bartekpacia resolved a small conflict due to a recently merged PR, can you review again? @isabelcosta can you review this too?Thank You

**Review by robotjellyzone — APPROVED** [review]

**Comment by bartekpacia** [comment]
@Techno-Disaster sure:D

**Review by bartekpacia — APPROVED** [review]

**Comment by isabelcosta** [comment]
thank you @HaripriyaB for testing this 🙌

**Review by isabelcosta — APPROVED** [review]
Thank you @Techno-Disaster for doing this :)

**Pull request merged**